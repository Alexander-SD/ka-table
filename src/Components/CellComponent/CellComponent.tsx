import * as React from 'react';

import { Cell } from '../../Models/Cell';
import { OptionChangedFunc } from '../../Types/OptionChangedFunc';
import {
  changeCellEditorToCellTextHandler, changeCellTextToCellEditorHandler,
} from '../../Utils/CellUtils';
import CellEditorState from '../CellEditorState/CellEditorState';
import CellText from '../CellText/CellText';

export interface ICellProps {
  field: string;
  onOptionChanged: OptionChangedFunc;
  rowData: any;
  rowKeyValue: any;
  isEditableCell: boolean;
  editableCells: Cell[];
}

const CellComponent: React.FunctionComponent<ICellProps> = ({
  field,
  rowKeyValue,
  rowData,
  isEditableCell,
  editableCells,
  onOptionChanged,
}) => {
  return (
    <td className='tc-cell'>
      { isEditableCell ? (
          <CellEditorState
            {...{ field, rowData }}
            onChangeToText={
              () => changeCellEditorToCellTextHandler({ field, rowKeyValue }, editableCells, onOptionChanged)
            }
          />
        )
        : (
          <CellText {...{ field, rowData }}
            onChangeToEditor={
              () => changeCellTextToCellEditorHandler({ field, rowKeyValue }, editableCells, onOptionChanged)
            }
          />
        )
      }
    </td>
  );
};

export default CellComponent;
