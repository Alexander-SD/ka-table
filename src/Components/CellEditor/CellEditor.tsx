import * as React from 'react';

import { DataType } from '../../Enums/DataType';
import { Column } from '../../Models/Column';
import { ValueChangeFunc } from '../../Types/ValueChangeFunction';
import CellEditorBoolean from '../CellEditorBoolean/CellEditorBoolean';
import CellEditorDate from '../CellEditorDate/CellEditorDate';
import CellEditorNumber from '../CellEditorNumber/CellEditorNumber';
import CellEditorString from '../CellEditorString/CellEditorString';

export interface ICellEditorProps {
  column: Column;
  rowData: any;
  onChangeToText: () => void;
  onValueChange: ValueChangeFunc;
}

const CellEditor: React.FunctionComponent<ICellEditorProps> = (props) => {
  const { dataType, editor } = props.column;
  if (editor) { return editor(props); }
  switch (dataType) {
    case DataType.Boolean: return <CellEditorBoolean {...props}/>;
    case DataType.Date: return <CellEditorDate {...props}/>;
    case DataType.Number: return <CellEditorNumber {...props}/>;
    default: return <CellEditorString {...props}/>;
  }
};

export default CellEditor;
