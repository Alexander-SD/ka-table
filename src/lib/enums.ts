export enum DataType {
  Boolean = 'boolean',
  Date = 'date',
  Number = 'number',
  Object = 'object',
  String = 'string',
}

export enum EditingMode {
  None = 'none',
  Cell = 'cell',
}

export enum ActionType {
  ChangeFilterRow = 'ChangeFilterRow',
  ChangeRowData = 'ChangeRowData',
  ChangeSorting = 'ChangeSorting',
  ChangeVirtualScrollingHeightSettings = 'ChangeVirtualScrollingHeightSettings',
  CloseEditor = 'CloseEditor',
  DeselectAllRows = 'DeselectAllRows',
  DeselectRow = 'DeselectRow',
  DeselectRowData = 'DeselectRowData',
  OpenEditor = 'OpenEditor',
  ScrollTable = 'ScrollTable',
  SelectAllRows = 'SelectAllRows',
  SelectSingleRow = 'SelectSingleRow',
  SelectRow = 'SelectRow',
  UpdateGroupsExpanded = 'UpdateGroupsExpanded',
}

export enum KeyboardEnum {
  Esc = 27,
  Enter = 13,
}

export enum SortDirection {
  Ascend = 'ascend',
  Descend = 'descend',
}

export enum SortingMode {
  None = 'none',
  Single = 'single',
}

export enum FilteringMode {
  None = 'none',
  FilterRow = 'filterRow',
}

export enum FilterOperatorName {
  Equal = '=',
  MoreThan = '>',
  LessThan = '<',
  MoreThanOrEqual = '>=',
  LessThanOrEqual = '<=',
  Contains = 'contains',
  IsEmpty = 'IsEmpty',
  IsNotEmpty = 'IsNotEmpty',
}
