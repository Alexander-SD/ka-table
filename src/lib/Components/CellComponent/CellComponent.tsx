import * as React from 'react';

import defaultOptions from '../../defaultOptions';
import { EditingMode } from '../../enums';
import { ChildAttributes } from '../../models';
import { Column } from '../../Models/Column';
import { DispatchFunc } from '../../types';
import { getField } from '../../Utils/ColumnUtils';
import { getValueByColumn } from '../../Utils/DataUtils';
import CellContent from '../CellContent/CellContent';
import CellEditor from '../CellEditor/CellEditor';

export interface ICellComponentProps {
  childAttributes: ChildAttributes;
  column: Column;
  dispatch: DispatchFunc;
  editingMode: EditingMode;
  isEditableCell: boolean;
  isSelectedRow: boolean;
  rowData: any;
  rowKeyField: string;
}

const CellComponent: React.FunctionComponent<ICellComponentProps> = (props) => {
  const {
    column,
    rowData,
    column: { style },
    isEditableCell,
  } = props;

  return (
    <td style={style} className={defaultOptions.css.cell}>
      { isEditableCell ? (
          <CellEditor
            {...props}
            value={getValueByColumn(rowData, column)}
            field={getField(column)}
          />
        )
        : (
          <CellContent
            {...props}
            value={getValueByColumn(rowData, column)}
            field={getField(column)}
          />
        )
      }
    </td>
  );
};

export default CellComponent;
