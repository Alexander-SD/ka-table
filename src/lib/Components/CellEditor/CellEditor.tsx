import * as React from 'react';

import defaultOptions from '../../defaultOptions';
import { EditingMode } from '../../enums';
import { ChildComponents } from '../../Models/ChildComponents';
import { Column } from '../../Models/Column';
import { DispatchFunc, Field, ValidationFunc } from '../../types';
import { getCellEditorDispatchHandler } from '../../Utils/CellUtils';
import { getElementCustomization } from '../../Utils/ComponentUtils';
import CellEditorState from '../CellEditorState/CellEditorState';

export interface IFilterRowEditorProps {
  childComponents: ChildComponents;
  column: Column;
  dispatch: DispatchFunc;
}

export interface ICellEditorProps extends IFilterRowEditorProps {
  autoFocus?: boolean;
  editingMode: EditingMode;
  editorValue?: any;
  field: Field;
  isDetailsRowShown: boolean;
  isSelectedRow: boolean;
  rowData: any;
  rowKeyField: string;
  rowKeyValue: any;
  value: any;
  validationMessage?: string;
  validation?: ValidationFunc;
}

const CellEditor: React.FunctionComponent<ICellEditorProps> = (props) => {
  const {
    childComponents,
    dispatch,
    editingMode,
    validationMessage
  } = props;

  const { elementAttributes, content } = getElementCustomization({
    className: `${defaultOptions.css.cellEditor} ${validationMessage ? defaultOptions.css.kaCellEditorValidationError : ''}`
  }, props, childComponents.editor);

  return (
    <div {...elementAttributes}>
      {content || (
        editingMode === EditingMode.Cell
        ? <CellEditorState {...props} dispatch={getCellEditorDispatchHandler(dispatch)} autoFocus={true}/>
        : <CellEditorState {...props} />
      )}
    </div>
  );
};

export default CellEditor;
