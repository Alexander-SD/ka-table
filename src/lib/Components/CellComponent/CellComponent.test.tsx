import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';

import { DataType, Events } from '../../enums';
import CellEditor from '../CellEditor/CellEditor';
import CellComponent from './CellComponent';

Enzyme.configure({ adapter: new Adapter() });
const props: any = {
  column: {
    dataType: DataType.String,
    field: 'columnField',
    title: 'Field',
  },
  editableCells: [],
  isEditableCell: false,
  onOptionChange: () => {},
  rowData: {
    column: 1,
  },
  rowKey: '1',
};

it('renders without crashing', () => {
  const element = document.createElement('tr');
  ReactDOM.render(<CellComponent {...props} />, element);
  ReactDOM.unmountComponentAtNode(element);
});

it('call RowDataChanged when onValueChange', () => {
  const dispatch = jest.fn();
  const newValue = {a: 1};
  const wrapper = mount(
    <CellComponent {...props} isEditableCell='true' dispatch={dispatch} />,
    {
      attachTo: document.createElement('tr'),
    },
  );
  wrapper.find(CellEditor).last().prop('onValueChange')(newValue);
  expect(dispatch).toBeCalledTimes(1);
  expect(dispatch).toBeCalledWith(Events.RowDataChanged, {newValue});
});
