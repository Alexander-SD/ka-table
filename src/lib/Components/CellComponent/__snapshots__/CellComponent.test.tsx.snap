// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call RowDataChanged when onValueChange 1`] = `
Array [
  Array [
    "RowDataChanged",
    Object {
      "newValue": Object {
        "a": 1,
      },
    },
  ],
]
`;
